agent/src/main/java/com/spotify/ffwd/statistics/SemanticCoreStatistics.java

```diff
@@ -246,24 +247,40 @@ public void reportMetricsDroppedByFilter(final int dropped) {
     @Override
     public HighFrequencyDetectorStatistics newHighFrequency() {
         final MetricId m = metric.tagged("component", "high-freq");
-        final AtomicLong highFreqMetrics = new AtomicLong();
+        final Map<MetricId, Long> highFreqMetricsMap = new ConcurrentHashMap<>();
 
         return new HighFrequencyDetectorStatistics() {
             private final Counter dropped = registry.counter(
                 m.tagged("what", "dropped-metrics", "unit", "metric"));
 
-            private final Gauge metricsCardinalityMetric =
-              registry.register(m.tagged("what", "high-freq-metrics"),
-                  (Gauge<Long>) () -> (long) highFreqMetrics.get());
-
             @Override
             public void reportHighFrequencyMetricsDropped(int dropped) {
                 this.dropped.inc(dropped);
             }
 
             @Override
-            public void reportHighFrequencyMetrics(int marked) {
-                highFreqMetrics.set(marked);
+            public void reportHighFrequencyMetrics(int marked, String... tags) {
+                MetricId gaugeMetric = m.tagged("what", "high-freq-metrics").tagged(tags);
+
+                registerGauge(gaugeMetric);
+
+                highFreqMetricsMap.put(gaugeMetric, (long) marked);
+            }
+
+            private void registerGauge(MetricId gauge) {
+                if (highFreqMetricsMap.containsKey(gauge)) {
```

Comment:
You can remove this now.  The entire logic is now atomic. You get an exception if you try to insert again.  
am I missing anything ?

Reply:
The register method should be called only in case this is new permutation of tags, otherwise every time it will throw the exception and I'd like to avoid it.